### DISCUSSION TITLE
Reusing vi.mock

### DISCUSSION THREAD

**Discussion by BibiSebi** [vitest-dev/vitest#4508] [Q&A] [answered]
Hello, I am currently working on a library and I would like to expose a setup() function that mocks all the necessary modules. So in the end the consumer of the library would use it somewhat like this: 

```

import { setup } from 'my-library'

 test('should test', async () => {
    setup()
    expect().toBe()
    .....
      })
```

Now the problem is, how can I make `vi.mock` reusable, so that it is evoked within setup(). Not sure if I am supposed to use something else here.

**Comment by sheremet-va** [comment]
`vi.mock` is statically analyzed and hoisted inside a file it is used in, so every import is affected. If you want to have it reusable, you would need to use dynamic import:

```ts
import { setup } from 'my-library'
setup() // use vi.doMock inside because it's not hoisted
const mod = await import('./mocked-module.js')
```

**Reply by BibiSebi** [reply]
Thank you for the quick response, i will give it a try :)

**Reply by BibiSebi** [reply]
@sheremet-va I was playing around with your suggested solution but I did not make it work. I am sending here my PR for a better overview of how we use the vi.mock. Maybe you even know some libraries that do something similar and where I can take an inspiration from? 

https://github.com/storyblok/field-plugin/pull/317/commits/5b0c4781bf35f4a4165a0fa059a9ba0dfb60bcff#diff-a18cfac33209a06a53f0d01ebc79f68d721f8782d0ed07697830dff36b9df73dR12

**Reply by sheremet-va** [reply]
I don't really understand what you are trying to achieve with this diff, to be honest. What is not working there?

**Reply by BibiSebi** [reply]
Oh ok, sorry, my mistake, I will try to explain it more clearly here. 

Context, our library has some internal functions that we would like to mock for testing. In the end the user should be able to call `setup()`function inside their test which will mock everything needed.

With vi.mock we are doing the following: 
```
vi.mock('@storyblok/field-plugin', async (importOriginal) => {
  const mod = await importOriginal<typeof import('@storyblok/field-plugin')>()

  const mockedCreateFieldPlugin: CreateFieldPlugin = ({ onUpdateState, validateContent }) =>
    mod.internalCreateFieldPlugin({
      onUpdateState,
      validateContent,
      postToContainer: (message) => {
        return mockContainer.onReceive(message)
      },
      listenToContainer: (handleMessage) => {
        mockContainer.setListener((message) => {
          return handleMessage(message)
        })
        return () => {
        }
      },
    })


  return {
    ...mod,
    createFieldPlugin: mockedCreateFieldPlugin,
  }
})
```

now I would like to extract vi.mock to a different file and pack it inside our `setup()` function. I tried using `vi.doMock` instead of `vi.mock`

and using dynamic import inside the test like this: 

```
import { setup } from 'my-library'

test('should increase the counter', async () => {
   await import('./mocked-module.js')
    setup()
  })
```

but no luck. I suspect the content of vi.doMock should also change somehow. Does this make more sense?

**Reply by sheremet-va** [reply, marked as answer]
> and using dynamic import inside the test like this:

This looks reversed. Setup should be called before the file that uses mocked import is imported. In your example `vi.doMock` is called after the module is imported. It should be something like this:

```ts
import { setup } from 'my-library'

test('should increase the counter', async () => {
   setup() // assuming that this calls "vi.doMock"
   await import('./mocked-module.js')
})
```

There is another solution to this, but it involves top-level await:

```ts
import { vi } from 'vitest'
import mocked from './mocked-module.js'

await vi.hoisted(async () => {
  // you can't have static import because this function is called before static imports are evaluated
  const { setup } = await import('my-library')
  setup() // assumes 'vi.doMock("./mocked-module.js")' is called
})

// "mocked" is now correctly mocked
```

**Reply by BibiSebi** [reply]
Thank you @sheremet-va.
